Override partial installs by default - part three

scheibelp opened the pull request:
Fixes #2794
Fixes #2634

This is intended to be a simpler approach with (most of) the same benefits as https://github.com/LLNL/spack/pull/3929. Compared to that PR, this PR has the disadvantage that reindex will mistakenly add partially-installed packages.

During install, remove prior unfinished installs

If a user performs an installation which fails, in some cases the
install prefix is still present, and the stage path may also be
present. With this commit, unless the user specifies
'--keep-prefix', installs are guaranteed to begin with a clean
slate. The database is used to decide whether an install finished,
since a database record is not added until the end of the install
process.

becker33 commented on lib/spack/spack/package.py:
```diff
@@ -1188,11 +1188,13 @@ def do_install(self,
         if self.spec.external:
             return self._process_external_package(explicit)
 
+        reuse_stage = not kwargs.get('restage', False)
```
code clarity note: Let's rename this to "restage" and flip its semantics. Then in self.repair_partial we can say `if restage: # do restage` and we avoid any double negatives

becker33 commented on lib/spack/spack/package.py:
```diff
@@ -1346,6 +1349,40 @@ def build_process(input_stream):
             if not keep_prefix:
                 self.remove_prefix()
 
+    def repair_partial(self, reuse_prefix=False, reuse_stage=True):
```
Can we rename this method to something more descriptive? Maybe "prepare_prefix_and_stage" or something like that? Also maybe rename "reuse_prefix" to "keep_prefix" and "reuse_stage" to "restage" (with the latter involving a value flip) so it's obvious to maintainers that this is managing the semantics of those two variables?

becker33 commented:
Please add a test for the restage option. This could be as simple as stage a package, touch a file in the stage, and install with keep-stage to see whether the file is still there.

scheibelp commented:
I renamed the mentioned variables and functions. IMO the existing test ```test_partial_install``` checks the ```restage``` argument. I renamed the test to ```test_partial_install_delete_prefix_and_stage```. Since the first review (but not yesterday's review) I added a ```MockStage``` object to confirm that the stage is destroyed when it should be. Let me know if this resolves your concerns and I'll rebase it.
